File: LeetCode/Java/src/solutions/p0121_BestTimetoBuyandSellStock/P0121.java
package solutions.p0121_BestTimetoBuyandSellStock;

public class P0121 {
    public int maxProfit(int[] prices) {

        // Brute Force
        // int max = 0;
        // for (int i = 0; i < prices.length; i++) {
        //     for (int j = i + 1; j < prices.length; j++) {
        //         int profit = prices[j] - prices[i];
        //         if (max < profit) {
        //             max = profit;
        //         }
        //     }
        // }
        // return max;

        // Runtime: 201 ms, faster than 6.83% of Java online submissions for Best Time to Buy and Sell Stock.
        // Memory Usage: 38.7 MB, less than 46.02% of Java online submissions for Best Time to Buy and Sell Stock.


        // 多加一步判断，但为什么更慢了？
        int max = 0, profit = 0;
        for (int i = 0; i < prices.length; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                if (prices[j] > prices[i]) {
                    profit = prices[j] - prices[i];
                }
                if (max < profit) {
                    max = profit;
                }
            }
        }
        return max;

        // Runtime: 347 ms, faster than 5.02% of Java online submissions for Best Time to Buy and Sell Stock.
        // Memory Usage: 38.8 MB, less than 30.53% of Java online submissions for Best Time to Buy and Sell Stock.
    }
}

File: LeetCode/Python/solutions/p0121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        #
        # Solution v1: Brute Force
        #
        # Runtime: 1069 ms @ (beats) 67.51%
        # Memory Usage: 25.1 MB @ (beats) 32.74%
        #

        ans = 0
        buy = float('inf')

        for p in prices:
            buy = min(buy, p)
            ans = max(ans, p - buy)

        return ans


if __name__ == "__main__":
    solution = Solution()

    prices = [7, 1, 5, 3, 6, 4]
    print(f"return: {solution.maxProfit(prices)}")  # 5

    prices = [7, 6, 4, 3, 1]
    print(f"return: {solution.maxProfit(prices)}")  # 0
